packages/solidity-contracts/contracts/messaging/gateway/FuelERC721Gateway.sol

```diff
@@ -0,0 +1,212 @@
+    /// @param fuelContractId ID of the contract on Fuel that manages the deposited tokens
+    /// @param tokenId tokenId to deposit
+    /// @dev Made payable to reduce gas costs
+    function deposit(
+        bytes32 to,
+        address tokenAddress,
+        bytes32 fuelContractId,
+        uint256 tokenId
+    ) external payable whenNotPaused {
+        bytes memory messageData = abi.encodePacked(
+            fuelContractId,
+            bytes32(uint256(uint160(tokenAddress))), // OFFSET_TOKEN_ADDRESS = 32
+            tokenId,
+            bytes32(uint256(uint160(msg.sender))), //from, OFFSET_FROM = 96
+            to, // OFFSET_TO = 128
+            uint256(1) // OFFSET_AMOUNT = 160
+        );
+        _deposit(tokenAddress, fuelContractId, tokenId, messageData);
+    }
+
+    /// @notice Deposits the given tokens to a contract on Fuel with optional data
+    /// @param to Fuel account or contract to deposit tokens to
+    /// @param tokenAddress Address of the token being transferred to Fuel
+    /// @param fuelContractId ID of the contract on Fuel that manages the deposited tokens
+    /// @param tokenId tokenId to deposit
+    /// @param data Optional data to send with the deposit
+    /// @dev Made payable to reduce gas costs
+    function depositWithData(
+        bytes32 to,
+        address tokenAddress,
+        bytes32 fuelContractId,
+        uint256 tokenId,
+        bytes calldata data
+    ) external payable whenNotPaused {
+        bytes memory messageData = abi.encodePacked(
+            fuelContractId,
+            bytes32(uint256(uint160(tokenAddress))), // OFFSET_TOKEN_ADDRESS = 32
+            tokenId, // OFFSET_TOKEN_ID = 64
+            bytes32(uint256(uint160(msg.sender))), //from, OFFSET_FROM = 96
+            to, // OFFSET_TO = 128
+            uint256(1), // OFFSET_AMOUNT = 160
+            DEPOSIT_TO_CONTRACT, // OFFSET_ROLE = 161
+            data
+        );
+        _deposit(tokenAddress, fuelContractId, tokenId, messageData);
+    }
+
+    /// @notice Finalizes the withdrawal process from the Fuel side gateway contract
+    /// @param to Account to send withdrawn tokens to
+    /// @param tokenAddress Address of the token being withdrawn from Fuel
+    /// @param tokenId Discriminator for ERC721 / ERC1155 tokens
+    /// @dev Made payable to reduce gas costs.
+    /// @dev Could remove the amount param to further reduce cost, but that implies changes in the Fuel contract
+    function finalizeWithdrawal(
+        address to,
+        address tokenAddress,
+        uint256 /*amount*/,
+        uint256 tokenId
+    ) external payable override whenNotPaused onlyFromPortal {
+        bytes32 fuelContractId = messageSender();
+        require(_deposits[tokenAddress][tokenId] == fuelContractId, "Fuel bridge does not own this token");
+
+        delete _deposits[tokenAddress][tokenId];
+
+        IERC721Upgradeable(tokenAddress).transferFrom(address(this), to, tokenId);
```

Comment:
Instantiating this instance each time to transfer is costly. Should we create a mapping of instances, fetch it from storage and then transfer?
That may or may not be cheaper over time.

Reply:
This is not an instantiation / creation. It is a call to another contract. Retrieving it from a mapping would be costlier. This has no gas cost to my knowledge.

In assembly code this would translate to `call(tokenAddress, some ugly params, area of memory where to store result, size of memory of result...)`